lib/f_core/protocols/horus/horus.c

```diff
@@ -0,0 +1,817 @@
+
+  5/ Compile for use as decoder called by  fsk_horus.m and fsk_horus_stream.m:
+
+    $ gcc horus_l2.c -o horus_l2 -Wall -DDEC_RX_BITS -DHORUS_L2_RX
+
+\*---------------------------------------------------------------------------*/
+
+#include "f_core/protocols/horus/horus.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define INTERLEAVER
+#define SCRAMBLER
+
+static char uw[] = {'$', '$'};
+
+/* Function Prototypes ------------------------------------------------*/
+
+int32_t get_syndrome(int32_t pattern);
+
+int golay23_init(void);
+
+int golay23_decode(int received_codeword);
+
+unsigned short gen_crc16(unsigned char *data_p, unsigned char length);
+
+void interleave(unsigned char *inout, int nbytes, int dir);
+
+void scramble(unsigned char *inout, int nbytes);
+
+/* Functions ----------------------------------------------------------*/
+
+/*
+   We are using a Golay (23,12) code which has a codeword 23 bits
+   long.  The tx packet format is:
+
+      | Unique Word | payload data bits | parity bits |
+
+   This function works out how much storage the caller of
+   horus_l2_encode_tx_packet() will need to store the tx packet
+ */
+
+int horus_l2_get_num_tx_data_bytes(int num_payload_data_bytes) {
+    int num_payload_data_bits, num_golay_codewords;
+    int num_tx_data_bits, num_tx_data_bytes;
+
+    num_payload_data_bits = num_payload_data_bytes * 8;
+    num_golay_codewords = num_payload_data_bits / 12;
+    if (num_payload_data_bits % 12) /* round up to 12 bits, may mean some unused bits */
+        num_golay_codewords++;
+
+    num_tx_data_bits = sizeof(uw) * 8 + num_payload_data_bits + num_golay_codewords * 11;
+    num_tx_data_bytes = num_tx_data_bits / 8;
+    if (num_tx_data_bits % 8) /* round up to nearest byte, may mean some unused bits */
+        num_tx_data_bytes++;
+
+    LOG_INF("num_payload_data_bytes: %d", num_payload_data_bytes);
+    LOG_INF("num_golay_codewords...: %d", num_golay_codewords);
+    LOG_INF("num_tx_data_bits......: %d", num_tx_data_bits);
+    LOG_INF("num_tx_data_bytes.....: %d", num_tx_data_bytes);
+
+    return num_tx_data_bytes;
+}
+
+/*
+  Takes an array of payload data bytes, prepends a unique word and appends
+  parity bits.
+
+  The encoder will run on the payload on a small 8-bit uC.  As we are
+  memory constrained so we do a lot of burrowing for bits out of
+  packed arrays, and don't use a LUT for Golay encoding.  Hopefully it
+  will run fast enough.  This was quite difficult to get going,
+  suspect there is a better way to write this.  Oh well, have to start
+  somewhere.
+ */
+
+int horus_l2_encode_tx_packet(unsigned char *output_tx_data, unsigned char *input_payload_data,
+                              int num_payload_data_bytes) {
+    int num_tx_data_bytes, num_payload_data_bits;
+    unsigned char *pout = output_tx_data;
+    int ninbit, ningolay, nparitybits;
+    int32_t ingolay, paritybyte, inbit, golayparity;
+    int ninbyte, shift, golayparitybit, i;
```

Comment:
Please >:(

Reply:
Just one line per variable. Pad out your GitHub stats. Also lets make sure things are initialized

Reply:
bro now were not c89 compatible